extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java

```diff
@@ -0,0 +1,76 @@
+package io.quarkus.infinispan.client.runtime.cache;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.commons.CacheException;
+
+import io.smallrye.mutiny.Uni;
+
+public class InfinispanGetWrapper {
+    final RemoteCache cache;
+    final Map<Object, ReentrantLock> synchronousGetLocks;
+
+    public InfinispanGetWrapper(RemoteCache cache, Map<Object, ReentrantLock> synchronousGetLocks) {
+        this.cache = cache;
+        this.synchronousGetLocks = synchronousGetLocks;
+    }
+
+    public <K, V> Uni<V> get(K key, Function<K, V> valueLoader) {
+        return Uni.createFrom().completionStage(
+                /*
+                 * Even if CompletionStage is eager, the Supplier used below guarantees that the cache value computation will be
+                 * delayed until subscription time. In other words, the cache value computation is done lazily.
+                 */
+                new Supplier<CompletionStage<V>>() {
+                    @Override
+                    public CompletionStage<V> get() {
+                        CompletionStage<Object> infinispanValue = getFromInfinispan(key, valueLoader);
+                        return cast(infinispanValue);
+                    }
+                });
+    }
+
+    private <K, V> CompletableFuture<Object> getFromInfinispan(K key, Function<K, V> valueLoader) {
+        ReentrantLock lock;
+        Object value = cache.get(key);
+        if (value == null) {
+            lock = synchronousGetLocks.computeIfAbsent(key, k -> new ReentrantLock());
```

Comment:
This code is blocking. We need to do something like instead

To do coalescing of operations I would suggest doing the following

```java
CompletionStage<Object> stage = new CompletionStage<>();
CompletionStage<Object> prev;
if ((prev = synchronousGetLocks.putIfAbsent(key, stage)) != null) {
   return prev;
}
cache.getAsync().whenComplete((v, t) -> {
   if (t != null) {
      stage.completeExceptionally(t);
   } else if (v != null) {
      stage.complete(v);
   } else {
      Object calculatedValue = ...
      if (calculatedValue != null) {
         cache.putIfAbsent(key, calculatedValue)
            .whenComplete((v, t) -> {
                if (t != null) {
                  stage.completeExceptionally(t);
                } else {
                  stage.complete(v == null ? calculatedValue : v);
                }
                synchronousGetLocks.remove(key);
            });
         return;
      }
   }
   synchronousGetLocks.remove(key);
});
return stage;

```

Reply:
To be 100% safe you may want to add a try/catch block around the entire thing and remove from `synchronousGetLocks` if any Throwable is thrown.